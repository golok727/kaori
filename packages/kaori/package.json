{
  "name": "kaori.js",
  "version": "0.0.4",
  "description": "A fragrant TS framework",
  "type": "module",
  "exports": {
    ".": "./src/index.ts",
    "./jsx-runtime": "./src/jsx/types.ts",
    "./jsx-dev-runtime": "./src/jsx/types.ts"
  },
  "publishConfig": {
    "exports": {
      ".": "./dist/index.js",
      "./jsx-runtime": "./dist/jsx/types.d.ts",
      "./jsx-dev-runtime": "./dist/jsx/types.d.ts"
    }
  },
  "files": [
    "dist",
    "src",
    "package.json"
  ],
  "scripts": {
    "build": "tsc"
  },
  "keywords": [
    "framework",
    "web",
    "frontend",
    "ui",
    "typescript",
    "kaori",
    "reactive",
    "signals",
    "lit-html"
  ],
  "author": "golok",
  "packageManager": "pnpm@10.12.4",
  "dependencies": {
    "@preact/signals-core": "^1.12.1",
    "lit-html": "^3.3.1"
  }
}
