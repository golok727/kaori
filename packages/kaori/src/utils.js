export function invariant(condition, message) {
    if (!condition) {
        throw new Error(message);
    }
}
//# sourceMappingURL=utils.js.map