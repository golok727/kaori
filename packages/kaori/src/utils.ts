export function invariant(
  condition: boolean,
  message: string
): asserts condition {
  if (!condition) {
    throw new Error(message);
  }
}

export const has_own = Object.hasOwn;
export const get_own_property_descriptor = Object.getOwnPropertyDescriptor;
export const define_property = Object.defineProperty;
