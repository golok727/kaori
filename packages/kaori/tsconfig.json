{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "./dist",
    "rootDir": "./src",
    "moduleResolution": "node",
    "tsBuildInfoFile": "./dist/tsconfig.tsbuildinfo"
  },
  "include": ["src", "src/jsx"]
}
