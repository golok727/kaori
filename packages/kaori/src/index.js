import { html } from "lit-html";
import { AsyncDirective, directive } from "lit-html/async-directive.js";
import { effect as syncEffect, untracked } from "@preact/signals-core";
import { invariant } from "./utils.js";
const logger = {
    log: (...args) => {
        console.log("[Kaori:DEV]", ...args);
    },
};
// todo move to a runtime file
/*
 Globals
 -----------------------------------------------------------------------
*/
let active_bloom = null;
let is_micro_task_queued = false;
let queued_microtasks = [];
let queued_updates = [];
let dirty_blooms = new WeakSet();
/*--------------------------------------------------------------- */
function queue_microtask(fn) {
    if (!is_micro_task_queued) {
        is_micro_task_queued = true;
        queueMicrotask(flush_microtasks);
    }
    if (fn) {
        queued_microtasks.push(fn);
    }
}
function flush_microtasks() {
    is_micro_task_queued = false;
    if (queued_microtasks.length > 0) {
        let microtasks = queued_microtasks;
        queued_microtasks = [];
        for (let i = 0; i < microtasks.length; i++) {
            microtasks[i]();
        }
    }
    const needs_updates = queued_updates;
    queued_updates = [];
    for (let i = 0; i < needs_updates.length; i++) {
        const [bloom, fn] = needs_updates[i];
        dirty_blooms.delete(bloom);
        fn();
    }
}
function schedule_update(bloom, fn) {
    if (dirty_blooms.has(bloom)) {
        return;
    }
    dirty_blooms.add(bloom);
    if (!is_micro_task_queued) {
        is_micro_task_queued = true;
        queueMicrotask(flush_microtasks);
    }
    queued_updates.push([bloom, fn]);
}
function on_mount(fn) {
    invariant(active_bloom !== null, "onMount() should be called during component setup (not in render functions or effects)");
    queue_microtask(fn);
}
function on_cleanup(fn) {
    const bloom = active_bloom;
    invariant(bloom !== null, "onCleanup() should be called during component setup (not in render functions or effects)");
    bloom.__disposables.add(fn);
}
function bloom_effect(fn, bloom = active_bloom) {
    invariant(bloom !== null, "effect() should be called during component setup (not in render functions or effects)");
    const dispose = untracked(() => syncEffect(fn));
    bloom.__disposables.add(dispose);
}
function get_bloom() {
    invariant(active_bloom !== null, "getBloom() can only be called during component setup (not in render functions or effects)");
    return active_bloom;
}
function dispose_bloom(bloom) {
    logger.log("Disposing bloom and running cleanup functions");
    bloom.__disposables.forEach((dispose) => {
        try {
            dispose();
        }
        catch (e) {
            console.error("Error during cleanup:", e);
        }
    });
    bloom.__disposables.clear();
}
class ComponentDirective extends AsyncDirective {
    constructor() {
        super(...arguments);
        this._rawTemplate = null;
        this.bloom = null;
        this.hasInitialized = false;
        this._cachedTemplate = null;
    }
    get $_getTemplate() {
        return typeof this._rawTemplate === "function"
            ? this._rawTemplate()
            : this._rawTemplate;
    }
    render(C, props) {
        const componentName = C.name || "Anonymous";
        if (!this.hasInitialized) {
            logger.log("Rendering component ", componentName, props);
            // Create the bloom object with update method
            this.bloom = {
                update: () => {
                    this._cachedTemplate = this.$_getTemplate;
                    if (this.bloom) {
                        schedule_update(this.bloom, () => {
                            if (this.isConnected) {
                                this.setValue(this._cachedTemplate);
                            }
                        });
                    }
                },
                __disposables: new Set(),
            };
            // Set current bloom context and call component
            const prevBloom = active_bloom;
            active_bloom = this.bloom;
            const result = untracked(() => C(props));
            active_bloom = prevBloom;
            this._rawTemplate = result;
            if (typeof this._rawTemplate === "function") {
                // reactive return
                bloom_effect(() => {
                    logger.log("Effect, ", componentName);
                    this.bloom?.update();
                }, this.bloom);
            }
            this._cachedTemplate = this.$_getTemplate;
            this.hasInitialized = true;
        }
        return this._cachedTemplate;
    }
    disconnected() {
        // Clean up the effect when component unmounts
        if (this.bloom) {
            dispose_bloom(this.bloom);
        }
    }
    reconnected() {
        // we need to do something here :w
        //
    }
}
export const component = directive(ComponentDirective);
export { For, Show } from "./helpers";
export { get_bloom as getBloom, bloom_effect as effect, on_mount as onMount, on_cleanup as onCleanup, html, };
export { render } from "lit-html";
export * from "./types";
export { signal, batch, computed, effect as syncEffect, untracked, } from "@preact/signals-core";
export { nothing } from "lit-html";
export { when } from "lit-html/directives/when.js";
export { choose } from "lit-html/directives/choose.js";
export { repeat } from "lit-html/directives/repeat.js";
export { map } from "lit-html/directives/map.js";
export { ref, createRef } from "lit-html/directives/ref.js";
//# sourceMappingURL=index.js.map