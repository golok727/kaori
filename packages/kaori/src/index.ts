import { html } from 'lit-html';
import { AsyncDirective, directive } from 'lit-html/async-directive.js';
import { effect as syncEffect, untracked } from '@preact/signals-core';
import type { Component } from './types.js';
import { invariant } from './utils.js';

export interface Logger {
  log: (...args: any[]) => void;
}

export const KaoriLogger: Logger = {
  log: (...args: any[]) => {
    console.log('[Kaori:DEV]', ...args);
  },
};

export const NoopLogger: Logger = {
  log: () => {
    // No operation
  },
};

let logger: Logger = NoopLogger;

export function _kaori_internal_set_logger_use_at_your_own_risk(
  logger: Logger
) {
  logger = logger;
}

// a handle to a component
export type ComponentHandle = {
  update(): void;
};

type ComponentHandleInternal = ComponentHandle & {
  __disposables: Set<() => void>;
  __dbg_n: string;
};

/*
 Globals
 -----------------------------------------------------------------------
*/
let active_handle: ComponentHandleInternal | null = null;
let is_micro_task_queued = false;
let queued_microtasks: Array<() => void> = [];
let queued_updates: Array<[ComponentHandleInternal, () => void]> = [];
let dirty_handles = new WeakSet<ComponentHandleInternal>();
/*--------------------------------------------------------------- */

function queue_microtask(fn: () => void) {
  if (!is_micro_task_queued) {
    is_micro_task_queued = true;
    queueMicrotask(flush_microtasks);
  }

  if (fn) {
    queued_microtasks.push(fn);
  }
}

function flush_microtasks() {
  is_micro_task_queued = false;

  if (queued_microtasks.length > 0) {
    let microtasks = queued_microtasks;
    queued_microtasks = [];
    for (let i = 0; i < microtasks.length; i++) {
      microtasks[i]();
    }
  }

  const needs_updates = queued_updates;
  queued_updates = [];

  for (let i = 0; i < needs_updates.length; i++) {
    const [handle, fn] = needs_updates[i];
    dirty_handles.delete(handle);
    fn();
  }
}

function schedule_update(handle: ComponentHandleInternal, fn: () => void) {
  if (dirty_handles.has(handle)) {
    return;
  }

  dirty_handles.add(handle);

  if (!is_micro_task_queued) {
    is_micro_task_queued = true;
    queueMicrotask(flush_microtasks);
  }

  queued_updates.push([handle, fn]);
}

function on_mount(fn: () => (() => void) | void) {
  invariant(
    active_handle !== null,
    'onMount() should be called during component setup (not in render functions or effects)'
  );

  const handle = active_handle;
  queue_microtask(() => {
    const cleanup = fn();
    if (typeof cleanup === 'function') {
      handle.__disposables.add(cleanup);
    }
  });
}

function on_cleanup(fn: () => void) {
  const handle = active_handle;
  invariant(
    handle !== null,
    'onCleanup() should be called during component setup (not in render functions or effects)'
  );
  handle.__disposables.add(fn);
}

function kaoi_effect(fn: () => void, handle = active_handle) {
  invariant(
    handle !== null,
    'effect() should be called during component setup (not in render functions or effects)'
  );

  queue_microtask(() => {
    const dispose = untracked(() => syncEffect(fn));
    handle.__disposables.add(dispose);
  });
}

function get_handle(): ComponentHandle {
  invariant(
    active_handle !== null,
    'getHandle() can only be called during component setup (not in render functions or effects)'
  );

  return active_handle;
}

function dispose_handle(handle: ComponentHandleInternal) {
  logger.log(`Disposing component(${handle.__dbg_n}) & running cleanup`);

  handle.__disposables.forEach(dispose => {
    try {
      dispose();
    } catch (e) {
      console.error('Error during cleanup:', e);
    }
  });
  handle.__disposables.clear();
}

class ComponentDirective<Props = any> extends AsyncDirective {
  private _rawTemplate: (() => unknown) | unknown = null;
  private handle: ComponentHandleInternal | null = null;
  private hasInitialized = false;

  private _cachedTemplate: unknown = null;

  private get $_getTemplate() {
    return typeof this._rawTemplate === 'function'
      ? this._rawTemplate()
      : this._rawTemplate;
  }

  render(C: Component<Props>, props: Props): unknown {
    const componentName = C.name || 'Anonymous';

    if (!this.hasInitialized) {
      logger.log('Initial render for component', componentName, props);
      this.handle = {
        __dbg_n: componentName,
        update: () => {
          this._cachedTemplate = this.$_getTemplate;
          if (this.handle) {
            schedule_update(this.handle, () => {
              if (this.isConnected) {
                this.setValue(this._cachedTemplate);
              }
            });
          }
        },
        __disposables: new Set<() => void>(),
      };

      const prev_handle = active_handle;
      active_handle = this.handle;
      const result = untracked(() => C(props));
      active_handle = prev_handle;

      this._rawTemplate = result;

      if (typeof this._rawTemplate === 'function') {
        // reactive return
        kaoi_effect(() => {
          logger.log('(effect) Update component: ', componentName);
          this.handle?.update();
        }, this.handle);
      }

      this._cachedTemplate = this.$_getTemplate;
      this.hasInitialized = true;
    }

    return this._cachedTemplate;
  }

  protected override disconnected(): void {
    // Clean up the effect when component unmounts
    if (this.handle) {
      dispose_handle(this.handle);
    }
  }

  protected override reconnected(): void {
    // we need to do something here :w
    //
  }
}

interface ComponentDirectiveFn {
  <Props>(C: Component<Props>, props: Props): unknown;
}

export const component = directive(ComponentDirective) as ComponentDirectiveFn;

export { For, Show } from './flow.js';
export { spread } from './helpers/spread.js';
export type { ForProps, ShowProps } from './flow.js';

export {
  /**
   * @deprecated use `getHandle` instead
   */
  get_handle as getBloom,
  get_handle as getHandle,
  kaoi_effect as effect,
  on_mount as onMount,
  on_cleanup as onCleanup,
  html,
};
export { render } from 'lit-html';

export * from './types';

export {
  signal,
  type Signal,
  type Effect,
  type Computed,
  batch,
  computed,
  effect as syncEffect,
  untracked,
} from '@preact/signals-core';

export { mergeProps } from './helpers/merge-props.js';
export { splitProps } from './helpers/split-props.js';
export { ifDefined } from 'lit-html/directives/if-defined.js';
export { nothing } from 'lit-html';
export { when } from 'lit-html/directives/when.js';
export { choose } from 'lit-html/directives/choose.js';
export { repeat } from 'lit-html/directives/repeat.js';
export { map } from 'lit-html/directives/map.js';
export { type Ref, ref, createRef } from 'lit-html/directives/ref.js';
export { styleMap } from 'lit-html/directives/style-map.js';
export { classMap } from 'lit-html/directives/class-map.js';
