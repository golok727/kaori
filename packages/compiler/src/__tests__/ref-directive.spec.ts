import { describe, it, expect } from 'vitest';
import { compile } from './utils';

describe('Ref directive transformation', () => {
  it('should transform basic ref usage', async () => {
    const input = `
import { createRef } from "kaori.js";

function App() {
  const myRef = createRef();
  return <div ref={myRef}>Hello</div>;
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });

  it('should transform ref with existing ref import', async () => {
    const input = `
import { ref, createRef, html } from "kaori.js";

function App() {
  const myRef = createRef();
  return <div ref={myRef}>Hello</div>;
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });

  it('should transform ref with conflicting name', async () => {
    const input = `
function App() {
  const ref = "some variable";
  const myRef = createRef();
  return <div ref={myRef}>Hello {ref}</div>;
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });

  it('should transform multiple refs in one component', async () => {
    const input = `
import { createRef } from "kaori.js";

function App() {
  const inputRef = createRef();
  const buttonRef = createRef();
  const divRef = createRef();

  return (
    <div ref={divRef}>
      <input ref={inputRef} type="text" />
      <button ref={buttonRef} onClick={handleClick}>
        Click me
      </button>
    </div>
  );
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });

  it('should transform ref with other attributes', async () => {
    const input = `
import { createRef } from "kaori.js";

function App() {
  const inputRef = createRef();

  return (
    <input
      ref={inputRef}
      type="text"
      className="input"
      prop:value={getValue()}
      bool:disabled={isDisabled()}
      onChange={handleChange}
      placeholder="Enter text"
    />
  );
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });

  it('should transform ref in nested elements', async () => {
    const input = `
import { createRef } from "kaori.js";

function App() {
  const outerRef = createRef();
  const innerRef = createRef();

  return (
    <div ref={outerRef} className="outer">
      <section>
        <p ref={innerRef}>Nested content</p>
      </section>
    </div>
  );
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });

  it('should transform ref with dynamic expression', async () => {
    const input = `
import { createRef } from "kaori.js";

function App() {
  const myRef = createRef();
  const shouldUseRef = true;

  return (
    <div ref={shouldUseRef ? myRef : undefined}>
      Conditional ref
    </div>
  );
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });

  it('should transform ref with inline createRef', async () => {
    const input = `
import { createRef } from "kaori.js";

function App() {
  return <div ref={createRef()}>Inline ref</div>;
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });

  it('should transform ref on component', async () => {
    const input = `
import { createRef } from "kaori.js";

function App() {
  const myRef = createRef();

  return (
    <>
      <MyComponent ref={myRef} />
      <div ref={myRef}>HTML element</div>
    </>
  );
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });

  it('should transform self-closing element with ref', async () => {
    const input = `
import { createRef } from "kaori.js";

function App() {
  const inputRef = createRef();
  const imgRef = createRef();

  return (
    <>
      <input ref={inputRef} type="text" />
      <img ref={imgRef} src="/logo.png" alt="Logo" />
    </>
  );
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });

  it('should transform ref with path import', async () => {
    const input = `
import { createRef, html, component } from "/@fs/C:/Users/Radha/dev/frameworks/kaori/packages/kaori/src/index.ts";

function App() {
  const myRef = createRef();
  return <div ref={myRef}>Hello</div>;
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });

  it('should transform aliased ref import', async () => {
    const input = `
import { ref as litRef, createRef } from "kaori.js";

function App() {
  const myRef = createRef();
  return <div ref={myRef}>Hello</div>;
}`;

    const output = await compile(input);
    expect(output).toMatchSnapshot();
  });
});
