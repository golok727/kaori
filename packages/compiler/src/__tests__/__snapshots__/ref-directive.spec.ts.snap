// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Ref directive transformation > should transform aliased ref import 1`] = `
"import { html } from "kaori.js";
import { ref as litRef, createRef } from "kaori.js";
function App() {
  const myRef = createRef();
  return html\`<div \${litRef(myRef)}>Hello</div>\`;
}"
`;

exports[`Ref directive transformation > should transform basic ref usage 1`] = `
"import { html, ref } from "kaori.js";
import { createRef } from "kaori.js";
function App() {
  const myRef = createRef();
  return html\`<div \${ref(myRef)}>Hello</div>\`;
}"
`;

exports[`Ref directive transformation > should transform multiple refs in one component 1`] = `
"import { html, ref } from "kaori.js";
import { createRef } from "kaori.js";
function App() {
  const inputRef = createRef();
  const buttonRef = createRef();
  const divRef = createRef();
  return html\`<div \${ref(divRef)}><input \${ref(inputRef)} type="text" /><button \${ref(buttonRef)} @click=\${handleClick}>
        Click me
      </button></div>\`;
}"
`;

exports[`Ref directive transformation > should transform ref in nested elements 1`] = `
"import { html, ref } from "kaori.js";
import { createRef } from "kaori.js";
function App() {
  const outerRef = createRef();
  const innerRef = createRef();
  return html\`<div \${ref(outerRef)} class="outer"><section><p \${ref(innerRef)}>Nested content</p></section></div>\`;
}"
`;

exports[`Ref directive transformation > should transform ref on component 1`] = `
"import { component, html, ref } from "kaori.js";
import { createRef } from "kaori.js";
function App() {
  const myRef = createRef();
  return [component(MyComponent, {
    ref: myRef
  }), html\`<div \${ref(myRef)}>HTML element</div>\`];
}"
`;

exports[`Ref directive transformation > should transform ref with conflicting name 1`] = `
"import { html, ref as ref1 } from "kaori.js";
function App() {
  const ref = "some variable";
  const myRef = createRef();
  return html\`<div \${ref1(myRef)}>Hello \${ref}</div>\`;
}"
`;

exports[`Ref directive transformation > should transform ref with dynamic expression 1`] = `
"import { html, ref } from "kaori.js";
import { createRef } from "kaori.js";
function App() {
  const myRef = createRef();
  const shouldUseRef = true;
  return html\`<div \${ref(shouldUseRef ? myRef : undefined)}>
      Conditional ref
    </div>\`;
}"
`;

exports[`Ref directive transformation > should transform ref with existing ref import 1`] = `
"import { ref, createRef, html } from "kaori.js";
function App() {
  const myRef = createRef();
  return html\`<div \${ref(myRef)}>Hello</div>\`;
}"
`;

exports[`Ref directive transformation > should transform ref with inline createRef 1`] = `
"import { html, ref } from "kaori.js";
import { createRef } from "kaori.js";
function App() {
  return html\`<div \${ref(createRef())}>Inline ref</div>\`;
}"
`;

exports[`Ref directive transformation > should transform ref with other attributes 1`] = `
"import { html, ref } from "kaori.js";
import { createRef } from "kaori.js";
function App() {
  const inputRef = createRef();
  return html\`<input \${ref(inputRef)} type="text" class="input" .value=\${getValue()} ?disabled=\${isDisabled()} @change=\${handleChange} placeholder="Enter text" />\`;
}"
`;

exports[`Ref directive transformation > should transform ref with path import 1`] = `
"import { ref } from "kaori.js";
import { createRef, html, component } from "/@fs/C:/Users/Radha/dev/frameworks/kaori/packages/kaori/src/index.ts";
function App() {
  const myRef = createRef();
  return html\`<div \${ref(myRef)}>Hello</div>\`;
}"
`;

exports[`Ref directive transformation > should transform self-closing element with ref 1`] = `
"import { html, ref } from "kaori.js";
import { createRef } from "kaori.js";
function App() {
  const inputRef = createRef();
  const imgRef = createRef();
  return [html\`<input \${ref(inputRef)} type="text" />\`, html\`<img \${ref(imgRef)} src="/logo.png" alt="Logo" />\`];
}"
`;
